app/src/components/common/TestURLModal/index.tsx

```diff
@@ -0,0 +1,228 @@
+
+  const timerRef = useRef(null);
+  const urlInputRef = useRef<InputRef>(null);
+  const sourceInputRef = useRef<InputRef>(null);
+
+  const renderResult = useCallback(() => {
+    if (!testURL) {
+      return (
+        <Row align="middle">
+          <InfoCircleOutlined />
+          <span className="result-text">
+            Match information will be displayed here automatically once you enter the URL.
+          </span>
+        </Row>
+      );
+    }
+
+    return isCheckPassed ? (
+      <Row align="middle">
+        <CheckCircleOutlined style={{ color: "var(--success)" }} />
+        <span className="result-text">
+          The URL <span style={{ color: "var(--success)" }}>matches</span> the source condition you defined.
+        </span>
+      </Row>
+    ) : (
+      <Row align="middle">
+        <CloseCircleOutlined style={{ color: "var(--danger)" }} />
+        <span className="result-text">
+          The URL <span style={{ color: "var(--danger)" }}> doesn't match </span> the source condition you defined.
+        </span>
+      </Row>
+    );
+  }, [testURL, isCheckPassed]);
+
+  const handleTestURL = (
+    url: string,
+    newValue: string = null,
+    newOperator: Source[keyof Source] = null,
+    newKey: Source[keyof Source] = null
+  ) => {
+    const config = {
+      value: newValue ?? sourceConfig.value,
+      operator: newOperator ?? sourceConfig.operator,
+      key: newKey ?? sourceConfig.key,
+    };
+    const result = RULE_PROCESSOR.RuleMatcher.matchUrlWithRuleSource(config, url, null);
+    setIsCheckPassed(result === "");
+  };
+
+  const handleSourceConfigChange = (key: keyof Source, value: Source[keyof Source]) => {
+    setSourceConfig((prevConfig) => ({ ...prevConfig, [key]: value }));
+    setIsSourceModified(true);
+  };
+
+  const handleSourceInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    clearTimeout(timerRef.current);
+    setSourceConfig((prevConfig) => ({ ...prevConfig, value: e.target.value }));
+    setIsSourceModified(true);
+    timerRef.current = setTimeout(() => {
+      handleTestURL(testURL, sourceInputRef.current.input.value);
+    }, 500);
+  };
+
+  const handleTestURLChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    clearTimeout(timerRef.current);
+    setTestURL(e.target.value);
+    setIsTestURLTried(true);
+    timerRef.current = setTimeout(() => {
+      handleTestURL(urlInputRef.current.input.value);
+    }, 500);
+  };
+
+  useEffect(() => {
+    trackURLConditionModalViewed(analyticsContext, source.operator);
+  }, [analyticsContext, source.operator]);
+
+  useEffect(() => {
+    if (isSourceModified) {
+      trackURLConditionSourceModified(analyticsContext, source.operator);
+    }
+  }, [analyticsContext, source.operator, isSourceModified]);
+
+  useEffect(() => {
+    if (isTestURLTried) {
+      trackURLConditionMatchingTried(analyticsContext, source.operator);
+    }
+  }, [isTestURLTried, analyticsContext, source.operator]);
```

Comment:
This can be moved in the debounced `handleTestURL`. This will get fired whenever a matching is tried